### ISSUE TITLE
`minimum_size_separators()` gives invalid result

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2521]
**Describe the bug**

`igraph_minimum_size_separators()` sometimes gives invalid results.

Note that the bug fix #2509 affects this function, but does not appear to fix this particular issue. This is expected, since this function executes its main algorithm only on connected graphs, and #2509 addresses disconnected graphs.

**To reproduce**

In Python,

```python
g=igraph.Graph([[0,8],[0,9],[1,2],[1,4],[2,4],[2,5],[3,4],[3,8],[5,6],[5,7],[6,9],[7,8]])
g.minimum_size_separators()
```

The result is:

```
[[2, 4],
 [8, 9],
 [6, 8],
 [5, 8],
 [3, 9],
 [3, 6],
 [3, 5],
 [4, 8, 9],
 [4, 6, 8],
 [4, 5],
 [4, 8],
 [5, 9],
 [0, 5],
 [0, 6],
 [2, 8],
 [2, 3]]
```

First, these obviously cannot all be minimum size separators, since some have a larger size than others. They can't even be minimal, since some are subsets of others (compare `[4,8,9]` and `[4,8]`).

Second, some of these are not separators at all. The graph looks like this:

<img width="360" alt="image" src="https://github.com/igraph/igraph/assets/1212871/d1e34498-daa4-4191-b261-c2751dfa280f">

The removal of `[3,9]` and `[3,6]` does not disconnect the graph.

_Note:_ This example was found by fuzzing, and the fuzzer couldn't minimize it further.

**Version information**
Current `master`.

**Comment by szhorvat** [comment]
Two more examples on 8 vertices, with 1-based vertex IDs:

```
{{2, 5}, {6, 7}, {5, 7}, {1, 6}, {4, 5}, {3, 4}, {4, 8}, {6, 8}, {1, 2}, {3, 8}}
```

<img width="360" alt="image" src="https://github.com/igraph/igraph/assets/1212871/8289e090-7c34-4c80-a163-07a77fbb5927">

```
{{2, 7}, {4, 8}, {3, 6}, {5, 8}, {5, 6}, {1, 2}, {2, 6}, {1, 5}, {3, 5}, {4, 7}, {5, 7}, {3, 4}, {1, 8}}
```

<img width="360" alt="image" src="https://github.com/igraph/igraph/assets/1212871/86c909dd-8825-41ec-ad73-ad92fbe14ba0">

**Comment by szhorvat** [comment]
The documentation of the `igraph_even_tarjan_reduction()` function states:

> For each original **directed** edge from \c i to \c j, two new edges are created

I think this should say "undirected edge" instead of "directed edge". `minimum_size_separators()` certainly passes undirected graphs to it. This should be corrected too when this is fixed.

(I noticed this earlier, I'm just writing down everything I had that's related to this issue.)

**Comment by vtraag** [comment]
I've had some time to look at it today. From what I gather from the publication, the Even-Tarjan reduction $\bar{G}_{st}$ is produced for a combination of source $s$ and target $t$.

![image](https://github.com/igraph/igraph/assets/6057804/5ca00be4-1a20-4eb1-84e2-5115a6e204f1)

However, in `igraph_even_tarjan_reduction`, this is not considered. I'm exploring whether reconstructing the Even-Tarjan graph during the for loop fixes the result.

**Comment by vtraag** [comment]
Note that the source and target node in the Even-Tarjan reduction do make a difference.

![image](https://github.com/igraph/igraph/assets/6057804/635f096c-8b56-4956-9068-ab3a72d266d7)

**Comment by vtraag** [comment]
What I don't fully understand is the following.

We rely on `igraph_all_st_mincuts` to find all minimum $s$-$t$ cuts in $\bar{G}_{st}$. This function returns the `stcuts`, the edges in the cutset in $\bar{G}_{st}$. However, what we need to return is sets of vertex separators (in $G$, not $\bar{G}_{st}$). But in `igraph_i_minimum_size_separators_append`, it seems that simply the edges in the cutset are being added to the `separators`.

It seems to me that we somehow need to transform the edges in the cutset of $\bar{G}_{st}$ into separating vertex sets of $G$. From Kanevsky (1993) it isn't directly obvious to me how this should be done. I'll look into it further some later time.

**Comment by vtraag** [comment]
Part of the problem seems to be that `igraph_all_st_mincuts` produce non-minimum cuts. I've opened a separate issue for that (issue #2532).

**Comment by szhorvat** [comment]
Fixed by #2533.
